patterns/self_encapsulation.md

```diff
@@ -0,0 +1,655 @@
+```ruby
+class Rectangle
+	def initialize(x: x, y: y, width: width, height: height)
+	end
+
+	def width
+	end
+
+	def height
+	end
+
+	def area
+	end
+
+	def perimeter
+	end
+end
+```
+When you define the method names, you are committing the class only to responding to those message names. You are not committing the class to any particular data structure or implementation, both of which could change in the future.
+
+You might not think that, in this example, the implementation of ```area``` would ever change for a rectangle, and that in the case of an immutable object, it would safe to simply assign the area as a property during construction. (Isn't ```area``` always ```width * height``` for a rectangle?) Recently on a project, we had to face just that assumption, when we realized that a 2D rectangle, if suspended in 3D not 2D space, actually had two kinds of 2D area: area when laid flat and area when viewed aerially (which would be equal to or less than the other area). Because we were more committed to the behavior of calculating area than the underlying data structure, we were able to make changes with minimal side effects.
+
+The point is, if you define a class by its behavior and not by its data structure, you minimize the upfront commitment and maximize the future possibilities and flexibility of the class.
+
+## Abstract State
+_**Also Known As: Implement Behavior with Abstract State**_
+
+### Context:
+
+You have identified the public behavior of a class (see Behavior-Defined Class above) but not yet implemented its behavior.
+
+### Problem:
+
+How do we approach implementation without prematurely and unnecessarily forcing data structure design decisions on subclasses?
+
+### Constraints:
+
+After you identifying the names and intents of protocols, your next step is to implement their behavior. For many method implementations, you will need access to the implied state of the object. For example, in order to calculate the area of a Circle, you need access to its radius.
+
+Many programmers' first instinct when they see the need for state is to define a data structure (e.g., ```this.radius = 10;``` in javascript). Don't do that! You can easily end up with a bunch of unnecessary state variables that you'd be better off deriving from existing variables or methods. For example, you can derive a Circle's diameter from its radius, so adding a second state variable for its diameter is unnecessary. Remember, every detail introduced into the data structure reduces the flexibility of subclass implementations.
+
+### Solution:
+
+**If you need implied state information in order to implement a behavior, define a message that returns that state instead of defining a variable.**
+
+For example, encapsulate the radius state inside a radius message. Don't call the state variable directly:
+
+```ruby
+class Circle
+	def radius # state-encapsulating method that returns a message
+	end
+
+	def pi # state-encapsulating method that returns a message
```

Comment:
Should probably lose this method as people will miss the point since `Math::PI` is the standard way to get `pi`